### PR TITLE
[ModelZoo] ESMM model

### PR DESCRIPTION
**Pull request by aalbersk**
Implementation of ESMM model based of EasyRec

### CODE REVIEW

**Comment by Duyi-Wang** [comment]
1. Please add model perf data in README, with reference to commit 9387a3406b3097990533f765caa835922b01a15f . 
2. Add distributed training code.
3. Remove .gitignore file

**Comment by app-on-mic** [comment]
Add more details of model requirements here.
Please follow the DLRM mode as the model template to finish the code, including the function of stand-alone and distributed training.
•	For stand-alone mode, please make sure the  performance and precision are ready on Alicloud instance.
•	For distributed mode, please make sure it run successfully in single machine following the steps below, which means the distributed training function is OK . (Please don’t care much about the distributed performance. Because the cluster functions depends on much effort and features which are not ready very well).  

# How to run distributed training in local

Set env by 

```shell
export TF_WORKSPACE=/home/DeepRec/modelzoo/ESMM # your model dir path
export TF_SCRIPT=train.py
```
use launch.py in DeepRec/modelzoo/DLRM/distribute_k8s/launch.py
Run the following three commods on three local bash.

```shell
TF_CONFIG='{"cluster": { "chief":["localhost:1111"], "ps":["localhost:2222"],"worker":["localhost:3333"]},"task":{"index":0,"type":"ps"}}' python launch.py --protocol=grpc --steps 2000 --output_dir=/tmp

TF_CONFIG='{"cluster": { "chief":["localhost:1111"], "ps":["localhost:2222"],"worker":["localhost:3333"]},"task":{"index":0,"type":"chief"}}' python launch.py --protocol=grpc --steps 2000 --output_dir=/tmp 

TF_CONFIG='{"cluster": { "chief":["localhost:1111"], "ps":["localhost:2222"],"worker":["localhost:3333"]},"task":{"index":0,"type":"worker"}}' python launch.py --protocol=grpc --steps 2000 --output_dir=/tmp
```

Please follow DLRM to add these switch:
- --inter
- --intra
- --protocol:just verify 'grpc' option, there are some bugs still not fixed in 'grpc++' & ‘star_sever’
- --input_layer_partitioner
- --dense_layer_partitioner

**Comment by Duyi-Wang** [comment]
Please use a 'head' scope to pack `self.compute_loss()`, acc & auc, it looks neater in tensorboard's graph, thx.
Please update community tensorflow version from 1.15 to 1.15.5 in DockerFile.

**Comment by aalbersk** [comment]
Thank you @app-on-mic for the instructions. They were very helpful.

@Duyi-Wang all of your comments are implemented.

**Comment by changqi1** [comment]
@aalbersk  please give us the following info:

image

**Comment by changqi1** [comment]
@aalbersk  I have 3 questions.
1)	For AUC, we should be more than 0.7, and use Ali-CCP dataset to improve the AUC.
2)	For Globalsteps/sec, why isn't DeepRec w/ oneDNN FP32+BF16 better than DeepRec w/ oneDNN FP32?  We want to get DeepRec w/ oneDNN FP32+BF16 better than DeepRec w/ oneDNN FP32.
3)	Need SPR data.

**Comment by Duyi-Wang** [comment]
@aalbersk acc calculation in line 212 need update.`self.acc, self.acc_op = tf.metrics.accuracy(labels=self.label, predictions=self.model)`
'tf.metrics.accuracy'  don't accept probability as predictions. 
Please update to ``self.acc, self.acc_op = tf.metrics.accuracy(labels=self.label, predictions=tf.round(self.model))``

File: modelzoo/ESMM/train.py [hunk 1]
```diff
@@ -0,0 +1,720 @@
...
+    test_file = os.path.join(args.data_location, 'taobao_test_data')
+    if not os.path.exists(args.data_location):
+        raise ValueError(f'[ERROR] data location: {args.data_location} does not exist. '
+                         'Please provide valid path')
+
+    if not os.path.exists(train_file) or not os.path.exists(test_file):
+        raise ValueError('[ERROR] taobao_train_data or taobao_test_data does not exist '
+                         'in the given data_location. Please provide valid path')
+
+    if not os.path.exists(args.output_dir):
```
Comment (Duyi-Wang) [inline modelzoo/ESMM/train.py:566]: Please remove the validation of `args.output_dir`. An error will occur in our regular test for Ali. If this directory does not exist, it will be automatically created.

Comment (aalbersk) [reply modelzoo/ESMM/train.py]: @Duyi-Wang done
